client/protocol.go

```diff
@@ -0,0 +1,413 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	curpapi "github.com/xline-kv/go-xline/api/curp"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+// Protocol client
+type protocolClient struct {
+	// Curp members
+	addrs []string
+	// All servers's `Connect`
+	connects map[uint64]string
+	// Curp client timeout settings
+	timeout clientTimeout
+}
+
+// Build client from addresses, this method will fetch all members from servers
+func BuildCurpClientFromAddrs(addrs []string) (*protocolClient, error) {
+	clientTimeout := NewDefaultClientTimeout()
+	cluster, err := fetchCluster(addrs, clientTimeout.proposeTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocolClient{
+		addrs:    addrs,
+		connects: cluster.AllMembers,
+		timeout:  clientTimeout,
+	}, nil
+}
+
+// Fetch cluster from server
+func fetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.FetchClusterResponse, error) {
+	var protocolClients []curpapi.ProtocolClient
+	fetchClusterRespCh := make(chan *curpapi.FetchClusterResponse)
+
+	for _, addr := range addrs {
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(1*time.Second))
```

Comment:
Please define a constant for the timeout.

Reply:
done